### PR TITLE
Revisit the metrics deletion flow

### PR DESCRIPTION
**Pull request by shabd67**
# Resolves

[CERES-515](https://jira.rax.io/browse/CERES-515)

# What

Revisit the metrics deletion flow

# How

https://github.com/Rackspace-Segment-Support/ceres-docs-internal/tree/master/proposals/delete-metric-flow

## How to test

Added JUnit Test Cases

### CODE REVIEW

**Review by rohitsngh27 — APPROVED** [review]
LGTM

File: src/main/java/com/rackspace/ceres/app/services/MetricDeletionService.java [hunk 1]
```diff
@@ -79,12 +81,21 @@ public MetricDeletionService(DataTablesStatements dataTablesStatements,
...
-        .then(deleteMetricsFromDownsampled(tenant, start, end))
-        .then(deleteMetricsFromRaw(tenant, start, end))
-        .then(Mono.empty());
+    Flux<String> seriesSetHashes = null;
+    if (start != null) {
+      return deleteMetricsFromDownsampled(tenant, start, end)
+          .then(deleteMetricsFromRaw(tenant, start, end))
+          .then(Mono.empty());
+    } else {
+      start = Instant.now().minus(getHighestTTLGranularity().getTtl());
```
Comment (matsboman) [inline src/main/java/com/rackspace/ceres/app/services/MetricDeletionService.java:90]: if start (and end) are not defined, shouldn't you delete all the data for this tenant id? Why are you setting a start value here? Also where are the metadata related to this tenant deleted?

File: src/main/java/com/rackspace/ceres/app/services/MetricDeletionService.java [hunk 2]
```diff
@@ -102,9 +113,18 @@ public MetricDeletionService(DataTablesStatements dataTablesStatements,
...
-        .then(metricDeletionHelper.deleteMetricNamesByTenantAndMetricName(tenant, metricName))
-        .then(Mono.empty());
+    if (start != null) {
+      final Instant startDateTime = start;
+      return seriesSetHashes.flatMap(seriesSetHash ->
+          deleteMetricsFromDownsampledWithSeriesSetHash(tenant, startDateTime, end, seriesSetHash)
+              .then(deleteMetricsFromRawWithSeriesSetHash(tenant, startDateTime, end, seriesSetHash)))
+              .then(Mono.empty());
+    } else  {
+      start = Instant.now().minus(getHighestTTLGranularity().getTtl());
```
Comment (matsboman) [inline src/main/java/com/rackspace/ceres/app/services/MetricDeletionService.java:123]: Same as previous comment

File: src/test/java/com/rackspace/ceres/app/services/MetricDeletionServiceTest.java [hunk 3]
```diff
@@ -122,11 +122,59 @@ public void testDeleteMetricsByTenantId() {
     metricDeletionService.deleteMetrics(tenantId, "", null,
         Instant.now().minusSeconds(60), Instant.now()).then().block();
 
-    assertViaQuery(tenantId, timeSlotPartitioner.rawTimeSlot(currentTime), seriesSetHash,
-        metricName);
+    //validate data raw
+    assertQueryRawViaQuery(tenantId, timeSlotPartitioner.rawTimeSlot(currentTime), 0);
```
Comment (matsboman) [inline src/test/java/com/rackspace/ceres/app/services/MetricDeletionServiceTest.java:126]: Are you asserting boundaries are deleted? E.g. start, end. Also in the middle.

File: src/test/java/com/rackspace/ceres/app/services/MetricDeletionServiceTest.java [hunk 4]
```diff
@@ -122,11 +122,59 @@ public void testDeleteMetricsByTenantId() {
...
+
+    //validate series_sets
+    assertSeriesSetViaQuery(tenantId, metricName, 4);
+
+    //validate metric_names
+    assertMetricNamesViaQuery(tenantId, metricName, 1);
+  }
+
+  @Test
+  public void testDeleteMetricsByTenantIdAndWithoutStartTime() {
```
Comment (matsboman) [inline src/test/java/com/rackspace/ceres/app/services/MetricDeletionServiceTest.java:139]: Here I would like to see that all the data is missing after a delete.

**Review by matsboman — COMMENTED** [review]
Some questions. It's not so clear to me what this is trying to do.

File: src/main/java/com/rackspace/ceres/app/services/MetricDeletionService.java (hunk 1 above)
Comment (shabd67) [reply src/main/java/com/rackspace/ceres/app/services/MetricDeletionService.java:90]: In order to delete all the metrics data along with its related metadata for that tenant, we need timeSlot and seriesSetHashes. So, at first I am querying all the seriesSetHashes to delete metadata in https://github.com/racker/ceres/blob/42d90969cbb4a8d45384acbe46d1c7c850e0911b/src/main/java/com/rackspace/ceres/app/services/MetricDeletionService.java#L94 and then deleting its metrics data afterwards.

File: src/test/java/com/rackspace/ceres/app/services/MetricDeletionServiceTest.java (hunk 4 above)
Comment (shabd67) [reply src/test/java/com/rackspace/ceres/app/services/MetricDeletionServiceTest.java:139]: yes, that's what is being checked as of now: https://github.com/racker/ceres/blob/42d90969cbb4a8d45384acbe46d1c7c850e0911b/src/test/java/com/rackspace/ceres/app/services/MetricDeletionServiceTest.java#L168

here, for all the tables in Cassandra expected rowCount is passed as 0 just to check all metric data along with its metadata is deleted

File: src/test/java/com/rackspace/ceres/app/services/MetricDeletionServiceTest.java (hunk 3 above)
Comment (shabd67) [reply src/test/java/com/rackspace/ceres/app/services/MetricDeletionServiceTest.java:126]: I am querying data from 60 seconds earlier to current time and onto them I am asserting data metrics data is deleted or not, keeping its metadata in its place.

File: src/main/java/com/rackspace/ceres/app/services/MetricDeletionService.java (hunk 1 above)
Comment (matsboman) [reply src/main/java/com/rackspace/ceres/app/services/MetricDeletionService.java:90]: `delete from <table> where tenant = 't-1';` 

Why not do like this?

File: src/test/java/com/rackspace/ceres/app/services/MetricDeletionServiceTest.java (hunk 3 above)
Comment (matsboman) [reply src/test/java/com/rackspace/ceres/app/services/MetricDeletionServiceTest.java:126]: I mean if you delete "from - to" shouldn't you check that from time stamp has no data anymore as well as to timestamp and also one sample in the middle? This is what I mean by boundary checking.

**Review by rohitsngh27 — APPROVED** [review]

**Review by matsboman — APPROVED** [review]
LGTM

**Pull request merged**